cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +131,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
+//
+// For those playing along that all means the first positional arg can be a
+// cloud/region, a region, a key=value to set, a key to get the settings for,
+// or empty. Other caveats are that one cannot set and reset a value for the
+// same key, that is to say keys to be mutated must be unique.
+//
+// Here we go...
 func (c *defaultsCommand) Init(args []string) error {
-	if c.reset {
-		// We're resetting defaults.
-		if len(args) == 0 {
-			return errors.New("no keys specified")
-		}
-		for _, k := range args {
+	var err error
+	//  If there's nothing to reset and no args we're returning everything. So
+	//  we short circuit immediately.
+	if len(args) == 0 && c.reset == "" {
+		c.action = c.getDefaults
+		return nil
+	}
+
+	// If there is an argument provided to reset, we turn it into a slice of
+	// strings and verify them. If there is one or more valid keys to reset and
+	// no other errors initalizing the command, c.resetDefaults will be called
+	// in c.Run.
+	if err = c.parseResetKeys(); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Look at the first positional arg and test to see if it is a valid
+	// optional specification of cloud/region or region. If it is then
+	// cloudName and regionName or only regionName are set on the object and
+	// the positional args are returned without the first element. If it cannot
+	// be validated; cloudName and regionName are left empty and we get back
+	// the same args we passed in.
+	args, err = c.parseArgsForRegion(args)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Remember we *might* have one less arg at this point if we chopped the
+	// first off because it was a valid cloud/region option.
+	switch {
+	case len(args) > 0 && strings.Contains(args[len(args)-1], "="):
+		// In the event that we are setting values, the final positional arg
+		// will always have an "=" in it. So if we see that we know we want to
+		// set args.
+		return c.handleSetArgs(args)
+	case len(args) == 0:
+		// This should be reset only (but possibly for a region).
+		return c.handleZeroArgs()
+	case len(args) == 1:
+		// We want to get settings for the provided key.
+		return c.handleOneArg(args[0])
+	default: // case args > 1
+		// Specifying any positional args after a key=value pair is
+		// invalid input. So if we have more than one the input is almost
+		// certainly invalid, but in different possible ways.
+		return c.handleExtraArgs(args)
+	}
+
+}
+
+// parseResetKeys splits the keys provided to --reset after trimming any
+// leading or trailing comma. It then verifies that we haven't incorrectly
+// received any key=value pairs and finally sets the value(s) on c.resetKeys.
+func (c *defaultsCommand) parseResetKeys() error {
+	if c.reset != "" {
+		resetKeys := strings.Split(strings.Trim(c.reset, ","), ",")
+
+		for _, k := range resetKeys {
 			if k == config.AgentVersionKey {
 				return errors.Errorf("%q cannot be reset", config.AgentVersionKey)
 			}
+			if strings.Contains(k, "=") {
```

Comment:
This is just trying to ensure they didn't type values to be set (a=b c=d) as comma separated to --reset. Validating the keys happens elsewhere. That is validating that we got things that look like keys, not that the keys are valid. Key validation happens elsewhere.